Core/src/EventData/CorrectedTransformationFreeToBound.cpp

```diff
@@ -0,0 +1,132 @@
+//
+// Copyright (C) 2021-2022 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/EventData/detail/CorrectedTransformationFreeToBound.hpp"
+
+#include "Acts/Surfaces/Surface.hpp"
+
+/// Get the non-linearity corrected bound parameters and its covariance
+std::optional<std::tuple<Acts::BoundVector, Acts::BoundSymMatrix>>
+Acts::detail::CorrectedFreeToBoundTransformer::operator()(
+    const Acts::FreeVector& freeParams,
+    const Acts::FreeSymMatrix& freeCovariance, const Acts::Surface& surface,
+    const Acts::GeometryContext& geoContext, NavigationDirection navDir) {
+  // Get the incidence angle
+  Vector3 dir = freeParams.segment<3>(eFreeDir0);
+  Vector3 normal = surface.normal(geoContext);
+  ActsScalar absCosIncidenceAng = std::abs(dir.dot(normal));
+  // No correction if the incidentAngle is small enough (not necessary ) or too
+  // large (correction could be invalid).
+  if (absCosIncidenceAng < cosIncidentAngleMinCutoff or
+      absCosIncidenceAng > cosIncidentAngleMaxCutoff) {
+    return std::nullopt;
+  }
+
+  size_t sampleSize = 2 * eFreeSize + 1;
+  std::vector<std::tuple<FreeVector, ActsScalar, ActsScalar>> sampledFreeParams;
+  sampledFreeParams.reserve(sampleSize);
+
+  // Initialize the covariance sqrt root matrix
+  FreeSymMatrix covSqrt = FreeSymMatrix::Zero();
+  // Get the covariance sqrt root matrix
+  Eigen::JacobiSVD<FreeSymMatrix> svd(
+      freeCovariance, Eigen::ComputeThinU | Eigen::ComputeThinV);
+  // U*S*V^-1
+  auto S = svd.singularValues();
+  auto U = svd.matrixU();
+  FreeMatrix D = FreeMatrix::Zero();
+  for (unsigned i = 0; i < eFreeSize; ++i) {
+    if (S(i) == 0) {
+      continue;
+    }
+    D(i, i) = std::sqrt(S(i));
+  }
+
+  FreeMatrix UP = FreeMatrix::Zero();
+  for (unsigned i = 0; i < eFreeSize; ++i) {
+    for (unsigned j = 0; j < eFreeSize; ++j) {
+      UP(i, j) = U(i, j);
+    }
+  }
+  covSqrt = UP * D;
+
+  double kappa = alpha * alpha * eFreeSize;
+  double gamma = std::sqrt(kappa);
+  double lambda = kappa - eFreeSize;
+
+  // Sample the free parameters
+  // 1. the baseline parameter
+  sampledFreeParams.push_back({freeParams, lambda / kappa,
+                               lambda / kappa + (1.0 - alpha * alpha + beta)});
+  // 2. the shifted parameters
+  for (unsigned i = 0; i < eFreeSize; ++i) {
+    sampledFreeParams.push_back(
+        {freeParams + covSqrt.col(i) * gamma, 0.5 / kappa, 0.5 / kappa});
+    sampledFreeParams.push_back(
+        {freeParams - covSqrt.col(i) * gamma, 0.5 / kappa, 0.5 / kappa});
+  }
+
+  // Initialize the mean of the bound parameters
+  BoundVector bpMean = BoundVector::Zero();
+  // Initialize the bound covariance
+  BoundSymMatrix bv = BoundSymMatrix::Zero();
+
+  // The transformed bound parameters
```

Comment:
That's transform bound + weight I suppose? Let's comment as such.

Reply:
Fixed.